File: src/ingest/core/record.py
"""Base class #1: Record (persistence as a class — no hand-written SQL).

A ClickHouse table is not a .sql file. It is a Record subclass that knows its
own columns, engine, and ordering, and GENERATES its own DDL. clickhouse/
migrate.py walks the Record subclasses and emits versioned migrations, so the
schema can never drift from the code.

Columns are declared as an ordered mapping of name -> ClickHouse type. Keeping
this dependency-free (no ORM) is deliberate: ClickHouse is analytical, append-
mostly, and the DDL is trivial to generate.
"""
from __future__ import annotations

from abc import ABC


class Record(ABC):
    __table__: str = ""
    __engine__: str = "MergeTree"
    __order_by__: tuple = ()
    __partition_by__: str | None = None
    #: ordered: column name -> ClickHouse type
    __columns__: dict = {}

    @classmethod
    def ddl(cls) -> str:
        if not cls.__table__ or not cls.__columns__:
            raise ValueError(f"{cls.__name__}: __table__ and __columns__ required")
        cols = ",\n  ".join(f"{n} {t}" for n, t in cls.__columns__.items())
        parts = [f"CREATE TABLE IF NOT EXISTS {cls.__table__} (\n  {cols}\n)"]
        parts.append(f"ENGINE = {cls.__engine__}")
        if cls.__partition_by__:
            parts.append(f"PARTITION BY {cls.__partition_by__}")
        if cls.__order_by__:
            parts.append(f"ORDER BY ({', '.join(cls.__order_by__)})")
        return "\n".join(parts) + ";"

    @classmethod
    def columns(cls) -> tuple:
        return tuple(cls.__columns__.keys())

    @classmethod
    def subclasses(cls) -> list:
        """All concrete Record tables — used by clickhouse/migrate.py."""
        out = []
        for sub in cls.__subclasses__():
            if sub.__table__:
                out.append(sub)
            out.extend(sub.subclasses())
        return out

File: internal/ledger/record.go
// Package ledger is base class #1 in Go form: a ClickHouse table is not a .sql
// file, it is a Record that knows its own columns, engine and ordering and
// generates its own DDL. cmd/chmigrate walks the registered Records and emits
// versioned migrations, so the schema can never drift from the code.
package ledger

import (
	"fmt"
	"strings"
)

// Column is one ClickHouse column. Declaration order is the column order, so
// Records carry a slice rather than a map.
type Column struct {
	Name string
	Type string
}

// Record is one table. Engine defaults to MergeTree; PartitionBy is optional.
type Record struct {
	Table       string
	Engine      string
	OrderBy     []string
	PartitionBy string
	Columns     []Column
}

// DDL renders the CREATE TABLE for this Record. It is deterministic: the same
// Record always renders byte-identical SQL, which is what lets the drift check
// compare generated output against committed migrations.
func (r Record) DDL() (string, error) {
	if r.Table == "" || len(r.Columns) == 0 {
		return "", fmt.Errorf("record %q: Table and Columns are required", r.Table)
	}

	cols := make([]string, len(r.Columns))
	for i, c := range r.Columns {
		if c.Name == "" || c.Type == "" {
			return "", fmt.Errorf("record %q: column %d has an empty name or type", r.Table, i)
		}
		cols[i] = fmt.Sprintf("  %s %s", c.Name, c.Type)
	}

	engine := r.Engine
	if engine == "" {
		engine = "MergeTree"
	}

	var b strings.Builder
	fmt.Fprintf(&b, "CREATE TABLE IF NOT EXISTS %s (\n%s\n)\n", r.Table, strings.Join(cols, ",\n"))
	fmt.Fprintf(&b, "ENGINE = %s\n", engine)
	if r.PartitionBy != "" {
		fmt.Fprintf(&b, "PARTITION BY %s\n", r.PartitionBy)
	}
	if len(r.OrderBy) > 0 {
		fmt.Fprintf(&b, "ORDER BY (%s)\n", strings.Join(r.OrderBy, ", "))
	}
	return strings.TrimRight(b.String(), "\n") + ";", nil
}

// registry holds every Record the binary knows about, in registration order.
var registry []Record

// Register adds a Record to the set the migrator walks. Call it from an init
// function in the file that declares the table.
func Register(r Record) {
	registry = append(registry, r)
}

// All returns the registered Records in declaration order.
func All() []Record {
	out := make([]Record, len(registry))
	copy(out, registry)
	return out
}
